src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -198,22 +203,182 @@ std::optional<Values> Visitor::visitTypesafe(Parser::ValuesClauseContext* ctx) {
+    return PropertyPath::fromVariable(var);
+  };
+  auto path = [](const PropertyPath& path) { return path; };
+  auto varOrPath = [&varPath,
+                    &path](ad_utility::sparql_types::VarOrPath varOrPath) {
+    return std::visit(ad_utility::OverloadCallOperator{varPath, path},
+                      std::move(varOrPath));
+  };
+
+  auto registerIfVariable = [this](auto& variant) {
+    if (holds_alternative<Variable>(variant)) {
+      visibleVariables_.back().emplace_back(std::get<Variable>(variant));
+    }
+  };
+
+  auto convertTriple = [&varOrTerm, &varOrPath, &registerIfVariable](
+                           TripleWithPropertyPath&& triple) -> SparqlTriple {
+    registerIfVariable(triple.subject_);
+    registerIfVariable(triple.predicate_);
+    registerIfVariable(triple.object_);
+
+    return {varOrTerm(std::move(triple.subject_)),
+            varOrPath(std::move(triple.predicate_)),
+            varOrTerm(std::move(triple.object_))};
+  };
+
+  vector<SparqlTriple> triples = ad_utility::transform(
+      ad_utility::flatten(
+          visitVector<vector<TripleWithPropertyPath>>(ctx->triplesBlock())),
+      convertTriple);
+  auto [ops, filters] =
+      splitUp(visitVector<std::variant<GraphPatternOperation, SparqlFilter>>(
+          ctx->graphPatternNotTriples()));
+  ops.emplace_back(
+      GraphPatternOperation::BasicGraphPattern{std::move(triples)});
+  return {ops, filters};
+}
+
 // ____________________________________________________________________________________
 vector<TripleWithPropertyPath> Visitor::visitTypesafe(
     Parser::TriplesBlockContext* ctx) {
   auto triples = visitTypesafe(ctx->triplesSameSubjectPath());
   if (ctx->triplesBlock()) {
-    appendVector(triples, visitTypesafe(ctx->triplesBlock()));
+    ad_utility::appendVector(triples, visitTypesafe(ctx->triplesBlock()));
   }
   return triples;
 }
 
+// ____________________________________________________________________________________
+std::variant<GraphPatternOperation, SparqlFilter> Visitor::visitTypesafe(
+    Parser::GraphPatternNotTriplesContext* ctx) {
+  // TODO: correctly set optional and id attributes
+  if (ctx->graphGraphPattern() || ctx->serviceGraphPattern()) {
+    throw ParseException(
+        "GraphGraphPattern or ServiceGraphPattern are not supported.");
+  } else if (ctx->filterR()) {
+    // TODO: visitAlternative
+    return visitTypesafe(ctx->filterR());
+  } else if (ctx->optionalGraphPattern()) {
+    auto optionalPattern = visitTypesafe(ctx->optionalGraphPattern());
+    optionalPattern._child._optional = true;
+    return GraphPatternOperation{std::move(optionalPattern)};
+  } else if (ctx->minusGraphPattern()) {
+    return GraphPatternOperation{visitTypesafe(ctx->minusGraphPattern())};
+  } else if (ctx->bind()) {
+    return GraphPatternOperation{visitTypesafe(ctx->bind())};
+  } else if (ctx->inlineData()) {
+    return GraphPatternOperation{visitTypesafe(ctx->inlineData())};
```

Comment:
I think using all the `GraphPatternOperation` cases for a `visitAlternative` abstraction, would even help us a lot.